### ISSUE TITLE
Linq mutator proposal

### ISSUE THREAD

**Issue by richardwerkman** [stryker-mutator/stryker-net#166]
I'm thinking about a cool new mutator for C#. There are a lot of LinQ methods that return the same interface, this makes them suitable for removing and switching. For example:

``` csharp
var query = someList.Where(x => x.Name == "Stryker").Distinct();
var query = someList.Where(x => x.Name == "Stryker"); // mutated
```
And
``` csharp
var query = someList.Where(x => x.Name == "Stryker").Reverse();
var query = someList.Where(x => x.Name == "Stryker"); // mutated
```
And
``` csharp
var query = someList.Where(x => x.Name == "Stryker").SingleOrDefault();
var query = someList.Where(x => x.Name == "Stryker").FirstOrDefault(); // mutated
```

What do you guys thinks about a mutator like this? Would this be useful, and if so, what kind of mutations should we make?

**Comment by Mobrockers** [comment]
There should be many fun things we can do with Linq queries, since most linq queries should return an IEnumerable. We should probably be careful that we do not create too many mutations since so much is possible, though. Your initial suggestions look good. 

I think especially the SingleOrDefault mutation is very useful.

**Comment by Valdas3** [comment]
Cool idea. I think that having mutations for All <-> Any, SkipWhile <-> TakeWhile, Min <-> Max might also be useful. Maybe even for OrderBy <-> OrderByDescending

**Comment by michielpeeters** [comment]
I'm not quite sure that you can actually break Linq statements. For example when someone is testing with one record in All() and Any() can result in a test which will survive and I think you want to break the tests all the time? And how do you handle when the Funcs are written for these linq statements? 

Are you searching specifically for a Where Linq statement + one of the above? Or do you want to handle multiples?

Sounds like a challenge :-)!

**Comment by Mobrockers** [comment]
A test surviving when any is switched with all would still indicate that your testset is inadequate, so I don't see the problem there. When your testset is adequate any and all cannot both be true.

**Comment by richardwerkman** [comment]
@Michielpeeters No the mutations are not specific for Linq statements with Where, that was just an example query. I'm not mutating Where in the Linq mutator with reason because the expression in the Where() will already be mutated by Stryker. An example on how this mutator will work is:

``` csharp
var query = someList.Distinct().OrderBy(x => x.Age);
```
would be mutated into
``` csharp
var query = someList.OrderBy(x => x.Age);
```
and 
``` csharp
var query = someList.Distinct();
```
When you are sure your tests are correct and still no test fails when removing `.Distinct()` for example, maybe the `.Distinct()` can be safely removed from your codebase. That way the mutation was still useful.

**Comment by richardwerkman** [comment]
Closed by creation of actual issue #182
